### PR TITLE
invoices: fix all invoices backlog delivery

### LINKED ISSUE #4311: Canceled invoices are not shown in SubscribeInvoice stream
### Background
Canceled invoices are not shown in SubscribeInvoice stream. 

Additionally node closes  socket and  `[ERR] INVC: unknown invoice state: Canceled` error appears in longs.

### Your environment

* version of `lnd`: v0.10.0-beta
* which operating system (`uname -a` on *Nix): Debian GNU/Linux 10 (buster)
* version of `btcd`, `bitcoind`, or other backend:  `"version": 190100,  "subversion": "/Satoshi:0.19.1/"`
* any other relevant environment details

### Steps to reproduce

0. Optionally use [Lightning Polar](https://lightningpolar.com/) - create a `Simnet` network with at least one LND node
1. Install Elixir. On macOS: `brew install elixir`, for Linux check distribution-specific step at https://elixir-lang.org/install.html#unix-and-unix-like
2. Clone and navigate into project folder `git@github.com:lessless/unistream.git && cd unistream`
3. Fetch dependencies mix deps.get
4. Edit TLS and Macaroon file locations if necessary in [lib/unistream.ex](https://github.com/lessless/unistream/blob/master/lib/unistream.ex)
5. Start app in a REPL shell `iex -S mix`
6. Run `Unistream.crashing_subscription` and observe error tuple `{:error, %GRPC.RPCError{message: ":stream_error: :closed", status: 13}}` 

**Or**

here is a JS scripts that will reproduce half of the problem - without dropping connection https://gist.github.com/lessless/f31eff5c07ef45f5d45c18a44c95af9c 

Be sure to edit path to cert and macaroon files.

### Expected behaviour

Canceled invoices should appear in stream and  server shouldn't close the socket.

### Actual behaviour

You will not see Canceled invoice in output stream and there will be ` [ERR] INVC: unknown invoice state: Canceled` error in node logs.
Also node closes socket.

### PR DESCRIPTION
**Pull request by joostjager**
Fixes #4311

### CODE REVIEW

**Comment by joostjager** [comment]
Maybe this isn't the right fix and it is better to deliver the real state for backlog deliveries. In the end, there is no way to straighten this bend without overhauling the whole subscription mechanism.

File: invoices/invoiceregistry.go [hunk 1]
```diff
@@ -420,6 +420,13 @@ func (i *InvoiceRegistry) deliverBacklogEvents(client *InvoiceSubscription) erro
 		// the loop reference causing is to point to the same item.
 		addEvent := addEvent
 
+		// Map the states that are unsupported in the all-invoices
+		// subscription to Open.
+		switch addEvent.State {
+		case channeldb.ContractAccepted, channeldb.ContractCanceled:
```
Comment (carlaKC) [inline invoices/invoiceregistry.go:426]: We should probably make a comment in the subscription rpc which explains this?

**Comment by joostjager** [comment]
In that case, what would the invoice state then be? Send an invoice with state Canceled into the Open channel? Or still remap the state to Open?

**Comment by carlaKC** [comment]
> Send an invoice with state Canceled into the Open channel

I'd say to send with it's state as is, because I think it would be worse to send an invoice with the wrong state than to send it signalling one state and then it's another. 

But there's no good way to do this without an overhaul. I guess the question is, which will be worse for existing systems using the api? Get invoices with the wrong state, or get invoices with unexpected state.

File: invoices/invoiceregistry_test.go [hunk 2]
```diff
@@ -947,3 +948,36 @@ func TestInvoiceExpiryWithRegistry(t *testing.T) {
...
+func TestSubscribeAll(t *testing.T) {
+	defer timeout()()
+
+	ctx := newTestContext(t)
+	defer ctx.cleanup()
+
+	// Add two invoices.
+	hash1 := lntypes.Hash{1}
+	_, err := ctx.registry.AddInvoice(testInvoice, hash1)
+	require.Nil(t, err)
```
Comment (bhandras) [inline invoices/invoiceregistry_test.go:962]: Nice to see `testify` making it's way to our tests :)
one nit: could use `assert.NoError(t, err)` here.

**Comment by joostjager** [comment]
>But there's no good way to do this without an overhaul. I guess the question is, which will be worse for existing systems using the api? Get invoices with the wrong state, or get invoices with unexpected state.

Or just don't get them at all, delete this PR?

File: invoices/invoiceregistry_test.go (hunk 2 above)
Comment (joostjager) [reply invoices/invoiceregistry_test.go]: fixed

**Comment by joostjager** [comment]
@cfromknecht what do you think about this PR? And.... close?

**Comment by cfromknecht** [comment]
Yeah i think it'll be better to address as part of the new invoice subscription rpc. Even though the logic doesn't do we _thought_ it did, it is still behavior that people may depend on in production.

**Comment by joostjager** [comment]
And close.

**Pull request closed without merging**